rhine/src/FRP/Rhine/Schedule.hs

```diff
@@ -17,30 +17,73 @@ and utilities to work with them.
 module FRP.Rhine.Schedule where
 
 -- base
-import Data.List.NonEmpty (NonEmpty (..))
-import Data.List.NonEmpty qualified as N
+import Control.Arrow
+import Data.List.NonEmpty as N
 
--- dunai
-import Data.MonadicStreamFunction
-import Data.MonadicStreamFunction.Async (concatS)
-import Data.MonadicStreamFunction.InternalCore
+-- transformers
+import Control.Monad.Trans.Reader
 
 -- monad-schedule
 import Control.Monad.Schedule.Class
 
 -- rhine
+import Data.Automaton
+import Data.Automaton.Final qualified as AutomatonFinal
+import Data.Automaton.MSF
+import Data.Automaton.Optimized (OptimizedAutomatonT (..), toAutomatonT)
+import Data.Automaton.Result
 import FRP.Rhine.Clock
 
 -- * Scheduling
 
 scheduleList :: (Monad m, MonadSchedule m) => NonEmpty (MSF m a b) -> MSF m a (NonEmpty b)
-scheduleList msfs = scheduleList' msfs []
+scheduleList msfs0 =
+  MSF $
+    Stateful $
+      AutomatonT
+        { state = (getFinal . toFinal <$> msfs0, [])
+        , step = \(msfs, running) -> ReaderT $ \a -> do
+            let bsAndConts = flip (runReaderT . AutomatonFinal.getFinal) a <$> msfs
+            (done, running') <- schedule (N.head bsAndConts :| N.tail bsAndConts ++ running)
+            return $ Result (resultState <$> done, running') $ output <$> done
+        }
+
+schedulePair :: (Monad m, MonadSchedule m) => MSF m a b -> MSF m a b -> MSF m a b
+schedulePair (MSF automatonL) (MSF automatonR) = MSF $! Stateful $! scheduleAutomata (toAutomatonT automatonL) (toAutomatonT automatonR)
   where
-    scheduleList' msfs running = MSF $ \a -> do
-      let bsAndConts = flip unMSF a <$> msfs
-      (done, running) <- schedule (N.head bsAndConts :| N.tail bsAndConts ++ running)
-      let (bs, dones) = N.unzip done
-      return (bs, scheduleList' dones running)
+    scheduleAutomata :: (Monad m, MonadSchedule m) => AutomatonT m b -> AutomatonT m b -> AutomatonT m b
+    scheduleAutomata (AutomatonT stateL0 stepL) (AutomatonT stateR0 stepR) =
+      AutomatonT
+        { state = (stepL stateL0, stepR stateR0)
+        , step
+        }
+      where
+        step (runningL, runningR) = do
+          result <- race runningL runningR
+          case result of
+            Left (Result stateL' b, runningR') -> return $ Result (stepL stateL', runningR') b
+            Right (runningL', Result stateR' b) -> return $ Result (runningL', stepR stateR') b
+
```

Comment:
Why two?

Reply:
scheduleList didn't work, although I'm not sure whether maybe some other bug was the issue back then